walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -24,86 +24,136 @@
 
 public final class FindMeetingQuery {
   
-  private Collection<TimeRange> timeOptions = new ArrayList<TimeRange>();
-  
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    timeOptions.clear();
+    
     if (request.getDuration() > TimeRange.WHOLE_DAY.duration()) {
       return Arrays.asList();
     }
     
-    if (request.getAttendees().isEmpty() || events.isEmpty()) {
+    Set<String> optionalRequestAttendees = buildAttendeeSet(request.getOptionalAttendees());
+    Set<String> mandatoryRequestAttendees = buildAttendeeSet(request.getAttendees());
+    Set<String> allRequestAttendees = combineAttendeeSets(optionalRequestAttendees, mandatoryRequestAttendees);
+    
+    if (allRequestAttendees.isEmpty() || events.isEmpty()) {
       return Arrays.asList(TimeRange.WHOLE_DAY);
     }
     
     Set<String> allEventAttendees = new HashSet<>();
     for (Event event : events) {
       allEventAttendees.addAll(event.getAttendees());
     }
-    Set<String> allRequestAttendees = new HashSet<>();
-    allRequestAttendees.addAll(request.getAttendees());
+    
+    int attendeesWithNoEvents = 0;
     for (String attendee : allRequestAttendees) {
       if (!allEventAttendees.contains(attendee)) {
-        return Arrays.asList(TimeRange.WHOLE_DAY);
+        attendeesWithNoEvents++;
       }
     }
+    if (attendeesWithNoEvents == allRequestAttendees.size()) {
+      return Arrays.asList(TimeRange.WHOLE_DAY);
+    }
+    
+    List<Event> allEvents = new ArrayList<Event>();
+    allEvents.addAll(events);
+    sortByStart(allEvents);
     
-    List<Event> eventsList = new ArrayList<Event>();
-    eventsList.addAll(events);
+    List<Event> optionalAttendeeEvents = new ArrayList<Event>();
+    List<Event> mandatoryAttendeeEvents = new ArrayList<Event>();
+    for (Event event : allEvents) {
+      if (optionalAttendeesOnly(event, mandatoryRequestAttendees)) {
+        optionalAttendeeEvents.add(event);
+      } else {
+        mandatoryAttendeeEvents.add(event);
+      }
+    }
+    sortByStart(optionalAttendeeEvents);
+    sortByStart(mandatoryAttendeeEvents);
+  
+    Collection<TimeRange> everyoneTimeOptions = getAllPossibleTimes(allEvents, request);
```

Comment:
Wouldn't it be more efficient to build this first and return it if it's not empty, which would avoid having to build mandatoryAttendeeEvents?

Reply:
Done